README
Color::Library::Dictionary::XKCD

The web comic "XKCD" <http://xkcd.com> ran a rather wide-spread color
survey to gauge the social perception of colors, at least, with regard
to screen print spaces.

The findings of this survey can be found here:
"blog.xkcd.com/2010/05/03/color-survey-results/"
<http://blog.xkcd.com/2010/05/03/color-survey-results/>

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Color-Library-Dictionary-XKCD

1.000002 2017-03-07T15:41:56Z 40dd0d5

1.000001 2017-03-07T15:41:17Z e43699f
 - Bugfix: Fix test failures with -Ddefault_inc_excludes_dot

 [Dependencies::Stats]
 - Dependencies changed since 1.000000, see misc/*.deps* for details
 - develop: +5 ↑2 -1 (suggests: ↑2)
 - test: ↓1

1.000000 2015-03-08T13:38:42Z cef6b43
 [Dependencies::Stats]
 - Dependencies changed since 0.1.0, see misc/*.deps* for details
 - build: -1
 - configure: +2 -1 (recommends: +1)
 - develop: +52 -2 (recommends: +1 -1, suggests: +1 ↑1)
 - runtime: +1 ↓1 (recommends: +1)
 - test: +4 ↓1 -2 (recommends: +3)

 [Features]
 - Colours now externalised to an independent file in a "share" directory to accelerate static analyisis tools on the
   module.

 [Meta]
 - Author is now @cpan.org

 [Packaging::Installer]
 - Installer is now EUMM
 - https://questhub.io/realm/perl/quest/5388cc12bbd0be5b2500011f

 [Packaging::Version]
 - Version scheme switched to x.yyyzzz
 - https://questhub.io/realm/perl/quest/5388d3a9bbd0be2224000129
 - This is the primary motivation for the 1.000000 on the box.

0.1.0 2013-07-04T13:56:06Z
 - First version.


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.032.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4211"
  },
  "configure_requires" => {
    "Module::Build" => "0.4211",
    "perl" => "5.006"
  },
  "dist_abstract" => "Colors determined by the XKCD Color Survey.",
  "dist_author" => [
    "Kent Fredric <kentfredric\@gmail.com>"
  ],
  "dist_name" => "Color-Library-Dictionary-XKCD",
  "dist_version" => "1.000000",
  "license" => "perl",
  "module_name" => "Color::Library::Dictionary::XKCD",
  "recommends" => {
    "Moo" => "1.000008"
  },
  "recursive_test_files" => 1,
  "requires" => {
    "Color::Library" => "0.021",
    "Color::Library::Dictionary" => 0,
    "Moo" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "1.301001_098",
    "perl" => "5.006"
  }
);


my %fallback_build_requires = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Module::Build" => "0.4211",
  "Test::More" => "1.301001_098",
  "perl" => "5.006"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


resources/assemble_colorlist.pl
#!/usr/bin/env perl

use strict;
use warnings;
use utf8;

use Path::Tiny qw( path );
use Data::Dumper qw( Dumper );

my @colors;

sub expand_color {
  my ( $title, $hexcolor ) = @_;
  my $name = $title;
  $name =~ s/\s+/_/g;
  my $id = "xkcd:$name";
  my ( $red, $green, $blue ) = map { hex $_ } grep { $_ } split /(..)/, $hexcolor;
  my $icolor = hex($hexcolor);
  return [ $id, $name, $title, [ $red, $green, $blue ], "$hexcolor", $icolor ];
}
my $index = 0;

for my $line ( path('.')->child('rgb.txt')->lines_utf8 ) {
  if ( $line =~ /^(.+?)\s+#([0-9a-f]{6})\s*$/ ) {
    my ( $title, $hexcolor ) = ( $1, $2 );
    push @colors, expand_color( $title, $hexcolor ), expand_color( "xkcd $index", $hexcolor );
    $index++;
    next;
  }
  die "$line doesn't match regex";
}

for my $color (@colors) {
  local $Data::Dumper::Terse     = 1;
  local $Data::Dumper::Purity    = 1;
  local $Data::Dumper::Useqq     = 0;
  local $Data::Dumper::Indent    = 0;
  local $Data::Dumper::Quotekeys = 0;
  my $line = Dumper($color);
  $line =~ s/\]$/,]/g;
  print qq[$line,\n];
}


resources/assemble_colorlist_pod.pl
#!/usr/bin/env perl 

use strict;
use warnings;
use utf8;

use Path::Tiny qw( path );
use Data::Dump qw( pp );

my @colors;

sub expand_color {
  my ( $title, $hexcolor ) = @_;
  my $name = $title;
  $name =~ s/\s+/_/g;
  my $id = "xkcd:$name";
  my ( $red, $green, $blue ) = map { hex $_ } grep { $_ } split /(..)/, $hexcolor;
  my $icolor = hex($hexcolor);
  return [ $id, $name, $title, [ $red, $green, $blue ], "$hexcolor", $icolor ];
}
my $index = 0;

my $labelids = {};

for my $line ( path('.')->child('rgb.txt')->lines_utf8 ) {
  if ( $line =~ /^(.+?)\s+#([0-9a-f]{6})\s*$/ ) {
    my ( $title, $hexcolor ) = ( $1, $2 );
    $labelids->{$title} = $index;
    push @colors, expand_color( $title, $hexcolor );

    #            expand_color( "xkcd $index", $hexcolor );
    $index++;
    next;
  }
  die "$line doesn't match regex";
}

for my $color ( sort { $a->[0] cmp $b->[0] } @colors ) {
  my $left = sprintf '%-25s -> %s ( xkcd_%-s )', $color->[2], $color->[1], $labelids->{ $color->[2] };
  printf "    %-70s -> #%s\n", $left, $color->[4];
}


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Colors determined by the XKCD Color Survey.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Color-Library-Dictionary-XKCD",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Color::Library::Dictionary::XKCD",
  "PREREQ_PM" => {
    "Color::Library" => "0.021",
    "Color::Library::Dictionary" => 0,
    "File::ShareDir" => 0,
    "Moo" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::File::ShareDir::Dist" => 0,
    "Test::More" => 0
  },
  "VERSION" => "1.000002",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Color::Library" => "0.021",
  "Color::Library::Dictionary" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::ShareDir" => 0,
  "File::Spec" => 0,
  "Moo" => 0,
  "Test::File::ShareDir::Dist" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
